### PR TITLE
Save some Cyborgs 

### PR DESCRIPTION
**Pull request by Fris0uman**
#### Summary
```SUMMARY: Features "Add the ability to turn prototype cyborg into NPC via Autodoc surgery"```

#### Purpose of change

Prototype Cyborgs are sad creatures that once were human but are now messed up robots.
Skilled survivor with enough heart should be able to remove the worse of the cbms messing them up to give them back some humanity.

#### Describe the solution

On (e)xamine Autodoc can now detect inactive broken cyborg or broken cyborg corpse waiting to revive.
Autodoc can't operate on broken cyborg, because they have too many machine parts.
Autodoc detect and can operate on inactive prototype cyborg
(If the corpse is too damaged Autodoc detect a dead patient)
 It will then prompt the player with the usual choice of in/uninstall.
Intsallation is impossible cyborgs are saturated in cbms
Removal offers some options, but most are unreadable due to the poor states of the patient.
Personality Override can be selected for removal
Upon succes an NPC of the Cyborg Class is spawned in place of the body
It has all the CBM Broken Cyborg profession has and the same gear.
Damage of the corpse item are apllied to the NPC
NPC spawn laying down
It will be very thankful to the  palyer and will offer to join them.

Chance of succes follow same logic as usual bionic removal
Damage on the cyborg body increases difficulty
On failure the body is damaged or destroyed

It's very strange but the Autodoc seems to systematically destroy the Personality Override cbm upon extraction...

Also this adds the ability to specify CBMs in classes.json

Prototype Cyborg is a new monster with : 
 - Same fighting ability as broken cyborg
 - Less defense than broken cyborg because they are more "fleshy"
 - some low cut damage because razor fingertips,
 - They are smart : 
                             - Can avoid traps
                             - Will prioritize targets
                             - Can open door : depending on #28734 
 - their own line of dialogue
 - same harvest group as broken cyborg
 - they will spawn in cells near operating theatre in labs, with a 50% chance to replace a broken cyborg
 - contrary to broken cyborg they can't revive
 - but they can be inactivated which is the only way ( for now ) to get them to the autodoc

#### Describe alternatives you've considered

There is no alternative, we can't let them suffer any longer ! Save the Cyborgs !
#### Additional context

This does not consume anesthetic kit, I could add it if it seems op without it. But I fear no one would use anesthetic kit on that.

I only wrote the bare minimum of background they could definitivly use more.
Personality Override CBM could have a better name, maybe a less obvious one.
I did not manage to set Cyborg gear up properly, so instead of subject suit and foot rags they only have subject suit.
I don't know if the faulty CBMs really affect the NPC in any way.
The ability to specify CBMs in classes.json works but might be a bit awkward.

### CODE REVIEW

**Comment by Fris0uman** [comment]
Crap I accidentally astyled all of game.cpp and iexamine.cpp

**Comment by SomeDeadGuy** [comment]
What? Broken cyborgs are dead human heads on robot parts. They was kept alive by electrical impulses in constant agony, cause AI need working human brain to operate, i guess.
The most humane thing you can do to a broken cyborg is to shut them down and never turn them on again or at least kill them with weapons.
i know, it is April Fools joke

**Comment by curstwist** [comment]
I hope it's not a joke but rather have a way to incapacitate them via laptop then work on corpses that dont revive as zombies. I feel so bad for the poor creatures

**Comment by Photoloss** [comment]
> What? Broken cyborgs are dead human heads on robot parts. They was kept alive by electrical impulses in constant agony, cause AI need working human brain to operate, i guess.

On the off-chance this is *not* a joke, broken cyborgs *do* have a few lines to the tune of "what have I become?!" and "please make it stop!". One line in particular ends with "self-preservation protocol activated" implying said program (mis?)identifies a noncompliant behaviour as attempting suicide which definitely counts as a complex sapient action. Of course they might still be beyond saving, either completely mad or taken over by the Blob.

Otherwise: [6879760](https://youtu.be/VXe9C3nZxxY?t=176). *I'm SO Sorry.*

**Comment by Wokko1** [comment]
Great idea, imo, why everyone thinks it's a joke?
Maybe make it only bruised corpses can be used on the autodoc, because when you rip them with your minigun, even if it will show the corpse as being "a little damaged", it still wouldn't make sense.

**Comment by aoae23** [comment]
Is there any way to make the procedure only work on alive broken cyborgs? That might make a bit more sense lore-wise than corpse revival.

**Comment by kevingranade** [comment]
Much like with "rescuing" zombies, this is impossible since the prerequisite for having the cyborg on the table is that you kill it first.  The autodoc doesn't have a way to bring someone back after death.

Even short of that, it's quite a reach that it would be capable of extracting key CBMs without killing the subject.

If you're interested in pursuing this, I suggest you have it simply extract CBMs from the corpse, and add the ability to revive the NPC to the Aftershock mod.

File: src/iexamine.cpp [hunk 1]
```diff
@@ -3906,15 +3922,56 @@ void iexamine::autodoc( player &p, const tripoint &examp )
...
+                    std::vector<std::string> choice_names;
+                    choice_names.push_back( "Personality_Override" );
+                    for( size_t i = 0; i < 6; i++ ) {
+                        choice_names.push_back( "C0RR#PTED?D#TA" );
+                    }
+                    int choice_index = uilist( _( "Choose bionic to uninstall" ), choice_names );
+                    if( choice_index == 0 ) {
+                        g->save_cyborg( &cyborg, couch_pos, p );
+                    } else {
+                        popup( _( "UNKNWON COMMAND.  Autodoc Mk. XI. Crashed." ) );
```
Comment (Rail-Runner) [inline src/iexamine.cpp]:
```suggestion
                        popup( _( "UNKNOWN COMMAND.  Autodoc Mk. XI. Crashed." ) );
```

File: data/json/npcs/TALK_CYBORG_1.json [hunk 2]
```diff
@@ -0,0 +1,11 @@
+[
+  {
+    "id": "TALK_CYBORG_1",
+    "type": "talk_topic",
+    "dynamic_line": "You... You saved me ! *Zzzt* Let me come with you, I can help!",
```
Comment (AMurkin) [inline data/json/npcs/TALK_CYBORG_1.json]:
```suggestion
    "dynamic_line": "You... You saved me!  *Zzzt* Let me come with you, I can help!",
```

File: data/json/npcs/Backgrounds/cyborg_1.json [hunk 3]
```diff
@@ -0,0 +1,9 @@
+[
+  {
+    "id": "BGSS_CYBORG_1_STORY1",
+    "type": "talk_topic",
+    "dynamic_line": "Oh God... *Zzzzt* I... I don't know what ha-happened.  *BEEEEEEP* They told me I was going to become the very b-best!  And then I c-can only remember pain and screams. P-Please don't *Zzzt* leave me here !",
```
Comment (AMurkin) [inline data/json/npcs/Backgrounds/cyborg_1.json]:
```suggestion
    "dynamic_line": "Oh God... *Zzzzt* I... I don't know what ha-happened.  *BEEEEEEP* They told me I was going to become the very b-best!  And then I c-can only remember pain and screams.  P-Please don't *Zzzt* leave me here!",
```

File: src/game.cpp [hunk 4]
```diff
@@ -5480,6 +5483,79 @@ bool game::revive_corpse( const tripoint &p, item &it )
...
+
+    if( !g->u.query_yn(
+            _( "WARNING: %i percent chance of SEVERE damage to all body parts! Continue anyway?" ),
+            ( 100 - static_cast<int>( chance_of_success ) ) ) ) {
+        return;
+    }
+
+    if( success > 0 ) {
+        add_msg( m_good, _( "Successfully removed Personality override." ) );
+        add_msg( m_bad, _( "Autodoc immediatly destroys the CBM uppon removal." ) );
```
Comment (AMurkin) [inline src/game.cpp]:
```suggestion
        add_msg( m_bad, _( "Autodoc immediately destroys the CBM upon removal." ) );
```

File: src/game.cpp [hunk 5]
```diff
@@ -5480,6 +5483,79 @@ bool game::revive_corpse( const tripoint &p, item &it )
...
+               dmg_lvl );
+
+        difficulty += dmg_lvl;// damage of the corpse increases difficulty
+    }
+
+    int chance_of_success = bionic_manip_cos( adjusted_skill, true, difficulty );
+    int success = chance_of_success - rng( 1, 100 ) ;
+
+    if( !g->u.query_yn(
+            _( "WARNING: %i percent chance of SEVERE damage to all body parts! Continue anyway?" ),
```
Comment (AMurkin) [inline src/game.cpp]:
```suggestion
            _( "WARNING: %i percent chance of SEVERE damage to all body parts!  Continue anyway?" ),
```

File: src/game.cpp [hunk 6]
```diff
@@ -5480,6 +5483,79 @@ bool game::revive_corpse( const tripoint &p, item &it )
...
+            ( 100 - static_cast<int>( chance_of_success ) ) ) ) {
```
Comment (AMurkin) [inline src/game.cpp]:
```suggestion
            100 - static_cast<int>( chance_of_success ) ) ) {
```

**Comment by Fris0uman** [comment]
> Maybe make it only bruised corpses can be used on the autodoc 

The damage to the body increases the difficulty and is transferred to the NPC but I could make that more stringent.

>Try to make "C0RR#PTED?D#TA" have randomly chosen "corrupted" characters. E.g. "C#RRUP?ED #ATA".

That's what I wanted to do but I don't know how.

>this is impossible since the prerequisite for having the cyborg on the table is that you kill it first.

Cyborg do come back to life after some time if you don't bash them, especially if you got them down with a EMP.
Putting inactive cyborg inactivated with control laptop on the autodoc couch also works. I could make that the only way to do it if it seems more appropriate.

>it's quite a reach that it would be capable of extracting key CBMs without killing the subject.

But it's already doing that in Npc and player so why not whith th humanoid cyborgs ? The difficulty can be raised further if you think it makes more sense.

>revive the NPC to the Aftershock mod.

If the active cyborg corpse that are going to revive are firmly considered dead I can remove that option and only make it possible to use inactive cyborg that are not dead.

**Comment by Fris0uman** [comment]
My idea was : 
Broken Cyborg have the REVIVE flag and the NO_NECRO one. Which means they can revive but can't be raised by Necromancer. 
 - So first they are not zombie which means ther's still hope for them.
 - And when killed without too much damage they are not completly dead since they are not zombies and will still wake up after some time.

At this point it's mostly a lore decision I guess. why can they revive ? What does it mean ? Do we consider them dead or not ?

**Comment by Rail-Runner** [comment]
Well, it does seem like a fun thing to do, in my opinion. Most of the cyborg's body seems to be metal, therefore the brain is likely kept alive by the body's artificial life-supporting systems, not by the blob (amount of flesh left is likely too small to sustain enough blob anyway). When we "kill" the cyborg, we render it unable to move, but life-supporting systems for the brain could possibly still be on (maybe get it to be randomly decided, something like 10% chance of it happening, otherwise it's brain-dead and you can't even attempt it).

**Comment by Mcrone617** [comment]
Another option to make it a bit lore friendly could be transforming the broken cyborgs into two enemies, the purely blob infected one, in which no humanity is left, and they are clearly dead (removal of human voice lines), and a much rarer proper cyborg that is still alive. Making it so only the control laptop shutting down the "alive" variant can lead to removal of the personality override CBM, and make it very hard to remove, and you have a pretty balanced reward to a lab, every fifth lab  or so you can, with great expense "create" a NPC.

I'm a huge fan of this idea, but if it is deemed sufficiently unrealistic, then I'm sure it at least belongs in Aftershock, it would be a shame to loose the potential to do this.

**Comment by int-ua** [comment]
I would love playing with this because I always play with NPC chance of .01 but not sure about enabling everything in Aftershock just for revival. No objections to making it really hard.

**Comment by paulenka-aleh** [comment]
I'm as well a huge fan of the idea even if it's going to be really hard, must be done properly via control laptop only (no damage, no EMP) and still has a chance to get a zombie-cyborg or evil cyborg back.

P.S. Oh, Personality Override!.. I already want this evil thing to be a proper CBM that I could try installing on my least favorite NPCs just for experiments :D

**Comment by kevingranade** [comment]
The rationale for the cyborg is that it has a bare minimum of human parts to operate (presumably because the creators could make *most* of a robot, but not some specific parts, usually including the brain).  As such, you cannot remove the cybernetic parts, because there is not enough human body left to make a viable person.

Assuming the cyborg is as it is because the original creators couldn't complete the cyborg, the only way out I can see for them is to take them to a (not yet existing) AI faction for repair and completion.  They would then become a "not broken cyborg", but it still wouldn't be an ordinary NPC, they'd be a member of the AI faction.

> But it's already doing that in Npc and player so why not whith th humanoid cyborgs ?

The difference is the extent of the modification, if we implement the proposed feature of "full cybernetic body", the regular autodoc wouldn't be able to reverse the process either.

> all of this comes from the Idea that somehow the Broken Cyborg profession is the same thing as the broken cyborg monster.

This is just a coincidence of names, they aren't the same at all.

I can think of a few ways forward with this concept:
1. As above, once we have the AI faction, they could fix the cyborg and make it a functional NPC, but it would be a member of the AI faction then.
2. A new partially cybernetic monster that IS just a NPC with a personality override.  Stats and such need to be distinct from the broken cyborg monster, pretty much start from scratch making a monster that represents a NPC with limited thinking ability, but some person-like abilities such as weapon use, and of course offensive CBMs. It wouldn't have the revives stuff, but it could be subject to shutdown with a control laptop and then you could drag it to an autodoc to try and remove the control module.

**Comment by Fris0uman** [comment]
Ok, so last option seems pretty good to me. I'm going to work on that.

**Comment by Fris0uman** [comment]
- New monster prototype cyborg : less armor and less dodge than broken cyborg but same skills since they were already low. They don't revive, they breath and can be inactivated. They have their on parrot lines but the harvest as broken cyborg
 - Autodoc still identifies Broken Cyborgs but can't operate because they are too machine and not enought human
- Only inactivated prototype cyborg can be operated by autodoc
- Prototype cyborg will spawn in lab, in the cells near operating room. I put them at the place that broken cyborg I don't know what are the chances to get one or the other.

File: data/json/items/tools.json [hunk 7]
```diff
@@ -902,6 +902,32 @@
       "skill2": "computer"
     }
   },
+  {
+    "id": "bot_prototype_cyborg",
+    "type": "TOOL",
+    "name": "inactive prototype cyborg",
+    "description": "This is a deactivated prototype cyborg, glimpses of humanity can still be seen in its empty eyes. Using this item involves placing it on the ground and reactivating its mechanical body.  If reprogrammed and rewired successfully the cyborg will then follow you and attack enemies. You monster.",
```
Comment (Night-Pryanik) [inline data/json/items/tools.json]:
```suggestion
    "description": "This is a deactivated prototype cyborg, glimpses of humanity can still be seen in its empty eyes.  Using this item involves placing it on the ground and reactivating its mechanical body.  If reprogrammed and rewired successfully the cyborg will then follow you and attack enemies.  You monster.",
```

File: data/json/monsters.json [hunk 8]
```diff
@@ -572,6 +572,37 @@
       "NO_NECRO"
     ]
   },
+  {
+    "id": "mon_prototype_cyborg",
+    "type": "MONSTER",
+    "name": "prototype cyborg",
+    "description": "A human fused with a mess of metal parts and wires.  While its eyes are empty, flashes of pain pass accross its face reminiscent of the person trapped in this grotesque body.  With enough surgical skills one might be able to give them back some humanity. If only they cared...",
```
Comment (Night-Pryanik) [inline data/json/monsters.json]:
```suggestion
    "description": "A human fused with a mess of metal parts and wires.  While its eyes are empty, flashes of pain pass across its face reminiscent of the person trapped in this grotesque body.  With enough surgical skills one might be able to give them back some humanity. If only they cared...",
```

File: data/json/speech.json [hunk 9]
```diff
@@ -1395,4 +1419,5 @@
         "sound" : "\"Police inbound.  Stay where you are!\"",
         "volume" : 15
     }
+
```
Comment (Night-Pryanik) [inline data/json/speech.json]:
```suggestion

```

**Comment by Photoloss** [comment]
> New monster prototype cyborg : less armor and less dodge than broken cyborg but same skills since they were already low.

Not sure why the existing broken cyborgs are so weak but maybe the "prototype" version could deal some (low) cut damage to mimic the fingertip razors CBM or exposed metal edges. Rare spawn with functioning (imitated) claws or blade bionic active? Going through the list of JSON flags you could additionally add `BONES` (set up harvest group to yield metal instead?), `CLIMBS` (better motor control than zombles), `ELECTRIC` (might be deprecated, alternatively `"special_when_hit": [ "ZAPBACK", 75 ]`), `GOODHEARING` (enhanced hearing CBM - good excuse for sensory overload torture as well), `GRABS` (might be a nerf with no supporting zombies) and/or `GUILT` (if not implied by `HUMAN`). Every bit of threat added also adds to the moral dilemma of trying to save them later. 

One **really** nasty option I don't see a flag for would be to allow them to *open* doors just like NPCs. This would make them *incredibly* dangerous to let loose without killing as you can't just lock them in the autodoc corridor. Since this would interfere with the integrity of tile separator doors it might be too much but as long as spawns are limited to the autodoc cells it would not lead to "bullshit RNG death" and all adverse effects would be entirely self-inflicted.

> Prototype cyborg will spawn in lab, in the cells near operating room. I put them at the place that broken cyborg I don't know what are the chances to get one or the other.

Test this, normally option-arrays can be given integer weights which would imply an equal split as the default. If weights don't work you can still limit it to a single cell.

**Comment by Fris0uman** [comment]
> Test this, normally option-arrays can be given integer weights which would imply an equal split as the default. If weights don't work you can still limit it to a single cell.

I left it as default so it should be 50/50 ? What's the synthaxe for weights ? [ [monster_a, 40], [monster_b,60] ] ?

> (low) cut damage to mimic the fingertip razors CBM

yep  I could add that.

> BONES (set up harvest group to yield metal instead?),

There supposed to be less machine than the broken cyborg so I don't think they should have metal bones.

>CLIMBS (better motor control than zombles), 

What good can climb do in the lab ? Also the broken cyborg profession has shitty motor control due to all the faulty bionics.

> zapback, goodhearing

They're supposed to be lower version of broken cyborg, wouldn't it be weird to give them more thant hte broken cyborgs ? And they will turn in Npc that won't have those ability.

>guilt

Guilt is always tricky, survivor don't get guilt when killing an aggressive NPC so it should be the same here.

>Opening doors

That could be a good idea. But they spawn in closed cells so I don't know if it will have much effect.

**Comment by Rail-Runner** [comment]
> What good can climb do in the lab ? Also the broken cyborg profession has shitty motor control due to all the faulty bionics.

> That could be a good idea. But they spawn in closed cells so I don't know if it will have much effect.

I think it should still be present if it makes sense, even if it's unlikely to come up under normal circumstances. Maybe someone decides to make a mod which will make them spawn in a location in which it does matter.

**Comment by Fris0uman** [comment]
Fair, but the CAN_OPEN_DOOR flag is not implemented yet.
And I don't get why it should have climb, I could be wrong but I think climb is a bit more than what a human can do and cyborg don't special feature that would give them climbing ability

**Comment by Photoloss** [comment]
> What's the synthaxe for weights ? [ [monster_a, 40], [monster_b,60] ] ?

I don't know the syntax in this case but the usual approach is to sum up the weights of all entries for the denominator and each individual entry as the numerator. So [40, 60] should yield the same results as [2, 3]. The "chance" entries are actual probability percentages where failed rolls yield none of the items, that's a different random draw concept.

> There supposed to be less machine than the broken cyborg so I don't think they should have metal bones.

Optional idea due to the skeletal bracing CBM, if you add the blade/claw variant limit it to those (military prototype Wolverine wannabe)

> What good can climb do in the lab ? Also the broken cyborg profession has shitty motor control due to all the faulty bionics.
> And I don't get why it should have CLIMB, I could be wrong but I think climb is a bit more than what a human can do and cyborg don't have special features that would give them climbing ability

"trap" rooms, maybe benefits similar to Parkour Expert or someone else might eventually add those later. Erk's aboveground facilities might provide more use-cases if you let them follow you upstairs. In case I am misunderstanding the flag the intent is to have them intelligently climb over chairs, tables and chainlink fences **not** to turn them into Spider-Man.

The profession still is a lot better off than zombies, more importantly to me these poor fellas should have the *cognitive* ability to recognise an obstacle and devise a smarter solution than headbutting into it.

> They're supposed to be lower version of broken cyborg, wouldn't it be weird to give them more thant hte broken cyborgs ? And they will turn in Npc that won't have those ability.

`ZAPBACK` can be a broken/harmful bionic removed during the rescue procedure. The hearing enhancement could be part of the personality override processor.

> Guilt is always tricky, survivor don't get guilt when killing an aggressive NPC so it should be the same here.

Definitely up for debate, my reasoning was that NPCs are just douchebags who *willingly* attack you while the cyborgs clearly are trapped in this state and *forced* to attack against their will.

> That could be a good idea. But they spawn in closed cells so I don't know if it will have much effect.

That is the beauty of it. The cells are locked so they can't do anything to you. But if you fiddle with the console and mess things up they will chase you all over the lab Terminator-style, especially if you add the enhanced hearing and maybe better night vision.

Basically the idea was to deliberately make them a fair bit **stronger** than "broken" cyborgs, greater risk for greater reward and all that, without just stat-creeping them in direct combat. While we're at it `PATH_AVOID_DANGER` (append `_integer`?) and `PRIORITIZE_TARGETS` also exist to enhance their supposed "intelligence".

**Comment by Fris0uman** [comment]
Looking at monmove code it seems that CLIMBS allow to move past some furniture AND climb up z levels.
I'm not sold on the zapback and enhanced hearing. But making them more intellignet is a good idea so I'll add that.
We'll have to make another PR later when open_doors is merged.

**Comment by Wokko1** [comment]
Open_doors is not merged until something in the game starts using it. It's [WIP] just because I still did not finish the mod, and thus no creature uses the flag yet.

If you want the open_door PR to get merged, probably you should make a PR using the flag, and then write something like "depends on #xxxxx", just so your cybors would justify adding the flag.

**Comment by Fris0uman** [comment]
I guess we can do it this way. But it might have been easier for you to add it to some monster like police bot for exemple within the PR itself.

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/guide-to-adding-new-content-to-cdda-for-first-time-modders/17781/31
